drivers/timer/apic_timer.c

```diff
@@ -0,0 +1,378 @@
+ *
+ * Even in TICKLESS mode, the timer is always counting. We maintain a soft
+ * 64-bit accumulator 'total_cycles' that holds the total number of cycles
+ * since we last started the timer, and that plus the number of cycles since
+ * timer restart (ICR - CCR) yields the total number of cycles since power-on.
+ *
+ * While 'total_cycles' is not, in general, bound to have a value that is
+ * an integral multiple of CYCLES_PER_TICK, it is a matter of design that the
+ * timer will always be programmed such that 'total_cycles' plus the timer
+ * count will result in an interrupt at a tick boundary. That is, (ICR +
+ * 'total_cycles') is always an integral multiple of CYCLES_PER_TICK.
+ */
+
+static void isr(void *arg); /* I love forward declarations in the morning */
+
+/*
+ * Only the very minimal bits of interest in the local APIC. Maybe for the
+ * sake of consistency these should be merged into include/drivers/loapic.h.
+ */
+
+#define DCR_DIVIDER_MASK	0x0000000F	/* divider bits */
+#define DCR_DIVIDER		0x0000000B	/* divide by 1 */
+#define LVT_MODE_MASK		0x00060000	/* timer mode bits */
+#define LVT_MODE		0x00000000	/* one-shot */
+
+/*
+ * CYCLES_PER_TICK must always be at least '2', otherwise MAX_TICKS
+ * will overflow s32_t, which is how 'ticks' are currently represented.
+ */
+
+#define CYCLES_PER_TICK \
+	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+BUILD_ASSERT_MSG(CYCLES_PER_TICK >= 2, "APIC timer: bad CYCLES_PER_TICK");
+
+/*
+ * MAX_TICKS is max number of ticks we can load into the timer in one shot.
+ */
+
+#define MAX_TICKS (0xFFFFFFFFU / CYCLES_PER_TICK)
+
+/*
+ * The spinlock protects all access to the local APIC timer
+ * registers, as well as 'total_cycles' and 'last_annoucement'.
+ */
+
+static struct k_spinlock lock;
+static u64_t total_cycles;	/* cycle accumulator */
+static u64_t last_announcement;	/* last time we called z_clock_announce() */
+
+/*
+ * The free-running timer referenced by z_timer_cycle_get_32() is, in an
+ * abstract sense, not required to be related to the timer driver proper,
+ * excepting that it must appear to have the same frequency. So, when
+ * appropriate, use the TSC frequency-scaled to provide this timer; when
+ * not available, use the decidedly inferior accumulator approximation.
+ */
+
+#ifdef CONFIG_APIC_TIMER_TSC
+
+u32_t z_timer_cycle_get_32(void)
+{
+	u64_t tsc = z_tsc_read();
+	u32_t cycles;
+
+	cycles = (tsc * CONFIG_APIC_TIMER_TSC_M) / CONFIG_APIC_TIMER_TSC_N;
+	return cycles;
+}
+
+#else
+
+/*
+ * As explained in the theory of operation, 'total_cycles' plus the cycles
+ * counted off is cycles-since-power on, subject to drift. We only return
+ * the lower 32 bits; we're obviously prepared if the API extends this to 64.
+ */
+
+u32_t z_timer_cycle_get_32(void)
+{
+	u32_t ret;
+	u32_t ccr;
+	u32_t icr;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	ccr = x86_read_loapic(LOAPIC_TIMER_CCR);
+	icr = x86_read_loapic(LOAPIC_TIMER_ICR);
```

Comment:
Sort of unfortunate naming.  The "ICR" register you'll find in the SDM is the "Interrupt Control Register".  This thing is the LVT Timer Initial Count, which doesn't seem to have a standard acronym.

On the subject of naming, and unrelated to this patch: we need to kill "loapic" dead.  That's a Zephyrism that no one else  understands.  The Intel term is just "APIC", or "Local APIC" if you absolutely have to disambiguate (but you never do, becuase the IO-APIC does a totally different job).  Linux uses "lapic", so I could see that.  But this name is weird.

Reply:
More meaty criticism: would be better to cache ICR rather than read it back from the hardware every time.   You're locked anyway, so it can only be a win.  That's what other drivers do with e.g. "last_load".

Reply:
That's an excellent point.  I will do that.

Reply:
And for the record, I didn't make any of `LOAPIC` or `ICR` here, these were pre-existing.  I agree that they're not the best choices.  I'll make a mental note to fix them after we retire the legacy local APIC timer driver.